### ISSUE TITLE
Replace autosend_email setting

### ISSUE THREAD

**Issue by hyanwong** [OneZoom/OZtree#480]
We previously declared an appconfig.ini variable called `smtp.autosend_email`, which we could set to 0 on the beta server, to make sure that we didn't accidentally automatically send emails on validation to sponsors.

However, it is actually quote useful to be able to test the emailing function, so I have currently coded it up so that if `is_testing` is set to "True" (which it normally is on beta), then any emails composed with the `ozmail.template_mail()` function are sent *to* the same email as they would normally appear *from*, which means sponsors shouldn't get accidental emails.

The question is, therefore, what to do with the `autosend_email` parameter. We *really* want to avoid sending emails to sponsors from anything but the production site. So even if is_testing is False, I think we would want to replace the send_to field in beta.

Therefore I think there should instead be an appconfig.ini variable called `email_recipient`. If Truthy (1, True, true, etc), we send to the normal recipient. If it is set to an email address, we use that as the recipient. If absent or Falsey, we do not send email at all.

I'm not sure if we actually want to bother checking `is_testing` after this. It just seems like an extra moving part.

We'd need to modify the unit tests etc to cope with this changed logic, but I don't think it would be too much work. What do you think @jrosindell  and @lentinj ?

**Comment by lentinj** [comment]
My preferred options here would be:

* Finish the job, move [the code in models/db.py](https://github.com/OneZoom/OZtree/blob/main/models/db.py#L144-L158) to [get_mailer()](https://github.com/OneZoom/OZtree/blob/main/modules/ozmail.py#L5-L24), make this the only source of mail objects.
* Add a ``smtp.force_sender`` option that makes ``get_mailer()`` return a wrapped mail object that sets the sender to the specified value (you might want to redirect the beta mail to your personal address temporarily, e.g.)
* Agree, leave ``is_testing`` out of it.

As it is ``smtp.autosend_email`` is a bit redundant too, you could empty the SMTP host option to achieve the same, but it's probably worth keeping for comfort.

**Comment by hyanwong** [comment]
Great comments, thanks @lentinj. Completely agree with getting all the code to use `get_mailer`. I think you mean `force_recipient` not `force_sender` though? the sender is always fixed as `smtp.sender` anyway.

I think we should get rid of `autosend_email` and replace it with some other variable like, as you say, `force_recipient`. Personally, I'd prefer that if no variable is defined in appconfig, no email is sent at all (or perhaps `smtp.sender` is forced as the recipient), when it's set to True, it works like on production (recipient as expected), and when it's an email address, it uses that to force the recipient. So it's just what *name* to use for this variable.

**Comment by lentinj** [comment]
>  I think you mean force_recipient not force_sender though?

Yes. Whoops :)

> replace it with some other variable like

Maybe just ``recipient``? So my config might look like:-

```
[smtp]
; Start a debug SMTP server: python -m smtpd -n -c DebuggingServer localhost:2500
server = localhost:2500
sender = me@example.com
tls = False
; Either:
; * "False" (default) - mail echoed to console
; * "mail@example.com" - all mail sent to given address
; * "True" - Mail sent to it's intended recipient
recipient = False
```

**Comment by hyanwong** [comment]
Yeah, "recipient" is good. Note, however, that we have to be careful not to flash the mail on the screen (if recipient=False), because we often have the same DB info on beta (where recipient could be `False`) and the mail contains private info, which could be revealed by simply typing in usernames or email addresses of other people.

**Comment by lentinj** [comment]
So how about:

* ``smtp.sender = None`` ==> Returned mail object is None, should fall back to putting mail in flash messages where relevant.
* ``smtp.recipient = False/None/Default`` ==> Get a mail object, but it just prints to stdout (i.e. server logs)
* ``smtp.recipient = "moo@example.com"`` ==> Get a mail object, but it doctors arguments to force recipient first
* ``smtp.recipient = True`` ==> Get a mail object with the safeties off.

**Comment by hyanwong** [comment]
LGTM

**Comment by hyanwong** [comment]
Did we fix this @lentinj ? There seems to be a bug whereby we are spamming donors (e.g. me!) with emails, even when `autosend_email = 0` in appconfig.ini. 

I think we need an urgent way to turn off the sending feature?

**Comment by davidebbo** [comment]
@hyanwong for what it's worth, I have not been getting any such emails. I got a single one on December 20th ("OneZoom sponsorship reminder"), and nothing since. I have both expired and active sponsorships.

**Comment by lentinj** [comment]
The quickest, most decisive way is turning off the background_tasks worker in supervisord.

**Comment by hyanwong** [comment]
Thanks Jamie. I wasn't sure of the syntax for turning off a specific program in supervisord (and not allowing it to restart), so I simply commented it out. If there is a command one can call, then it might be useful to document it in the conf file.

Can I confirm that as long as this task/program is off, I can set `autosend_email` back on again without risking spamming other donors?

**Comment by hyanwong** [comment]
> supervisorctl stop [job]

I assume the only time this will then restart is if the entire supervisord process is restarted (or we issue a `supervisorctl start www.onezoom.org_background_tasks`)?

> Any other e-mail we send will be part of a web request, so it should only be things they trigger.

Great, I'll turn autosend_email=1 back on then.

**Comment by lentinj** [comment]
> I assume the only time this will then restart is if the entire supervisord process is restarted 

Correct. You could turn off [``autostart``](http://supervisord.org/configuration.html#program-x-section-values) to stop this happening if you wanted.
